horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/UserServiceImpl.java

```diff
@@ -1,66 +1,34 @@
 package io.hyperfoil.tools.horreum.svc;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
-import java.util.stream.Collectors;
-
-import io.quarkus.arc.properties.IfBuildProperty;
+import io.hyperfoil.tools.horreum.api.internal.services.UserService;
+import io.hyperfoil.tools.horreum.entity.user.UserInfo;
+import io.hyperfoil.tools.horreum.server.WithRoles;
+import io.hyperfoil.tools.horreum.svc.user.UserBackEnd;
 import io.quarkus.security.Authenticated;
+import io.quarkus.security.identity.SecurityIdentity;
 import jakarta.annotation.security.PermitAll;
 import jakarta.annotation.security.RolesAllowed;
 import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.inject.Instance;
 import jakarta.inject.Inject;
 import jakarta.transaction.Transactional;
-import jakarta.ws.rs.ClientErrorException;
-import jakarta.ws.rs.NotFoundException;
-import jakarta.ws.rs.WebApplicationException;
-import jakarta.ws.rs.core.Response;
-
-import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.jboss.logging.Logger;
-import org.keycloak.admin.client.Keycloak;
-import org.keycloak.admin.client.resource.RoleMappingResource;
-import org.keycloak.admin.client.resource.RoleResource;
-import org.keycloak.admin.client.resource.RoleScopeResource;
-import org.keycloak.representations.idm.ClientRepresentation;
-import org.keycloak.representations.idm.CredentialRepresentation;
-import org.keycloak.representations.idm.RoleRepresentation;
 import org.keycloak.representations.idm.UserRepresentation;
 
-import io.hyperfoil.tools.horreum.api.internal.services.UserService;
-import io.hyperfoil.tools.horreum.entity.user.UserInfo;
-import io.hyperfoil.tools.horreum.server.WithRoles;
-import io.quarkus.security.identity.SecurityIdentity;
-import io.vertx.core.Vertx;
+import java.util.List;
+import java.util.Map;
 
 @PermitAll
 @ApplicationScoped
-@IfBuildProperty(name = "horreum.roles.provider", stringValue = "keycloak", enableIfMissing = true)
 public class UserServiceImpl implements UserService {
    private static final Logger log = Logger.getLogger(UserServiceImpl.class);
-   private static final String[] ROLE_TYPES = new String[] { "team", Roles.VIEWER, Roles.TESTER, Roles.UPLOADER, Roles.MANAGER };
-
-   @ConfigProperty(name="horreum.keycloak.realm", defaultValue="horreum")
-   String realm;
-
-   @Inject
-   Keycloak keycloak;
-
    @Inject
    SecurityIdentity identity;
 
    @Inject
-   Vertx vertx;
+   Instance<UserBackEnd> backend;
 
-   private static UserData toUserInfo(UserRepresentation rep) {
+   public static UserData toUserInfo(UserRepresentation rep) {
```

Comment:
this `toUserInfo` implementation can be moved to keycloak backend and kept private

Reply:
Have reinstated in KeycloakUserBackend